File: python/lexer/lexer_test.py
import unittest

from lexer import Lexer
import tokens as tok
from tokens import Token


class Test(unittest.TestCase):
    def test_get_next_token_1(self):
        input = "=+(){},;"

        tokens = [
            tok.ASSING,
            tok.PLUS,
            tok.LPAREN,
            tok.RPAREN,
            tok.LSQUIRLY,
            tok.RSQUIRLY,
            tok.COMMA,
            tok.SEMICOLON,
        ]

        lex = Lexer(input)
        for token in tokens:
            self.assertEqual(lex.get_next_token().type, token)

    def test_get_next_token_2(self):
        input = """\
let five = 5;
let ten = 10;
let add = fn(x, y) {
    x + y;
};
let result = add(five, ten);
!-/*5;
5 < 10 > 5;
if (5 < 10) {
return true;
} else {
return false;
}
"""
        lex = Lexer(input)
        tokens = [
            Token(tok.LET, "let"),
            Token(tok.IDENT, "five"),
            Token(tok.ASSING, "="),
            Token(tok.INT, "5"),
            Token(tok.SEMICOLON, ";"),
            Token(tok.LET, "let"),
            Token(tok.IDENT, "ten"),
            Token(tok.ASSING, "="),
            Token(tok.INT, "10"),
            Token(tok.SEMICOLON, ";"),
            Token(tok.LET, "let"),
            Token(tok.IDENT, "add"),
            Token(tok.ASSING, "="),
            Token(tok.FUNCTION, "fn"),
            Token(tok.LPAREN, "("),
            Token(tok.IDENT, "x"),
            Token(tok.COMMA, ","),
            Token(tok.IDENT, "y"),
            Token(tok.RPAREN, ")"),
            Token(tok.LSQUIRLY, "{"),
            Token(tok.IDENT, "x"),
            Token(tok.PLUS, "+"),
            Token(tok.IDENT, "y"),
            Token(tok.SEMICOLON, ";"),
            Token(tok.RSQUIRLY, "}"),
            Token(tok.SEMICOLON, ";"),
            Token(tok.LET, "let"),
            Token(tok.IDENT, "result"),
            Token(tok.ASSING, "="),
            Token(tok.IDENT, "add"),
            Token(tok.LPAREN, "("),
            Token(tok.IDENT, "five"),
            Token(tok.COMMA, ","),
            Token(tok.IDENT, "ten"),
            Token(tok.RPAREN, ")"),
            Token(tok.SEMICOLON, ";"),
            Token(tok.BANG, "!"),
            Token(tok.MINUS, "-"),
            Token(tok.SLASH, "/"),
            Token(tok.ASTERISK, "*"),
            Token(tok.INT, "5"),
            Token(tok.SEMICOLON, ";"),
            Token(tok.INT, "5"),
            Token(tok.LESSTHAN, "<"),
            Token(tok.INT, "10"),
            Token(tok.GREATERTHAN, ">"),
            Token(tok.INT, "5"),
            Token(tok.SEMICOLON, ";"),
            Token(tok.IF, "if"),
            Token(tok.LPAREN, "("),
            Token(tok.INT, "5"),
            Token(tok.LESSTHAN, "<"),
            Token(tok.INT, "10"),
            Token(tok.RPAREN, ")"),
            Token(tok.LSQUIRLY, "{"),
            Token(tok.RETURN, "return"),
            Token(tok.TRUE, "true"),
            Token(tok.SEMICOLON, ";"),
            Token(tok.RSQUIRLY, "}"),
            Token(tok.ELSE, "else"),
            Token(tok.LSQUIRLY, "{"),
            Token(tok.RETURN, "return"),
            Token(tok.FALSE, "false"),
            Token(tok.SEMICOLON, ";"),
            Token(tok.RSQUIRLY, "}"),
            Token(tok.EOF, "EOF"),
        ]
        for token in tokens:
            self.assertEqual(lex.get_next_token(), token)

File: lua/lexer/lexer_test.lua
#!/usr/bin/env luajit

local lex = require ".lexer"

local test_single_char_tokens = function()
  local input = "=+(){},;"
  lex.lexer(input)
  local expected = {
    lex.token_t.ASSIGN,
    lex.token_t.PLUS,
    lex.token_t.LPAREN,
    lex.token_t.RPAREN,
    lex.token_t.LSQUIRLY,
    lex.token_t.RSQUIRLY,
    lex.token_t.COMMA,
    lex.token_t.SEMICOLON,
    lex.token_t.EOF,
  }
  for _, t in ipairs(expected) do
    local g = lex.next_token().type
    if g ~= t then
      error(t .. "Not matching as expected", 1)
    end
  end
end

local test_multiple_char_tokens = function()
  local input = "let five = 5;\n"
    .. "let ten = 10;\n"
    .. "let add = fn(x, y) {\n"
    .. "     x + y;\n"
    .. "};\n"
    .. "let result = add(five, ten);"
  lex.lexer(input)
  local expected = {
    { lex.token_t.LET, "let" },
    { lex.token_t.IDENTIFIER, "five" },
    { lex.token_t.ASSIGN, "=" },
    { lex.token_t.INTEGER, "5" },
    { lex.token_t.SEMICOLON, ";" },
    { lex.token_t.LET, "let" },
    { lex.token_t.IDENTIFIER, "ten" },
    { lex.token_t.ASSIGN, "=" },
    { lex.token_t.INTEGER, "10" },
    { lex.token_t.SEMICOLON, ";" },
    { lex.token_t.LET, "let" },
    { lex.token_t.IDENTIFIER, "add" },
    { lex.token_t.ASSIGN, "=" },
    { lex.token_t.FUNCTION, "fn" },
    { lex.token_t.LPAREN, "(" },
    { lex.token_t.IDENTIFIER, "x" },
    { lex.token_t.COMMA, "," },
    { lex.token_t.IDENTIFIER, "y" },
    { lex.token_t.RPAREN, ")" },
    { lex.token_t.LSQUIRLY, "{" },
    { lex.token_t.IDENTIFIER, "x" },
    { lex.token_t.PLUS, "+" },
    { lex.token_t.IDENTIFIER, "y" },
    { lex.token_t.SEMICOLON, ";" },
    { lex.token_t.RSQUIRLY, "}" },
    { lex.token_t.SEMICOLON, ";" },
    { lex.token_t.LET, "let" },
    { lex.token_t.IDENTIFIER, "result" },
    { lex.token_t.ASSIGN, "=" },
    { lex.token_t.IDENTIFIER, "add" },
    { lex.token_t.LPAREN, "(" },
    { lex.token_t.IDENTIFIER, "five" },
    { lex.token_t.COMMA, "," },
    { lex.token_t.IDENTIFIER, "ten" },
    { lex.token_t.RPAREN, ")" },
    { lex.token_t.SEMICOLON, ";" },
    { lex.token_t.EOF, "eof" },
  }
  for _, t in ipairs(expected) do
    local g = lex.next_token()
    if g.literal ~= t[2] or g.type ~= t[1] then
      error(t[2] .. "Not matching as expected", 1)
    end
  end
end

test_single_char_tokens()
test_multiple_char_tokens()
